refactor internal codes (to_kana)

JSYoo5B opened the pull request:
On resolving bugfix about InputContext, I'm doubting `to_kana()` method.

In order to fix bugs in InputContext, I think refactoring is necessary for debugging.
No offence, but it's difficult for me to understand code flow with the current codebase.

I'm just trying to refactor internal codes, not the public interfaces & constants.
Detailed reasons for its refactoring were described in each commit log.
I suppose this refactor may not finish very soon. So I'm marking it with WIP

JSYoo5B committed:
refactor(automata): rename hangul & kana consts

* Every consts has `HANJP_` prefix, confuses its usage
  * Rename Hangul Jaso prefix into `HANGUL_`
    (reverts 68a8a8fd004d359d3ffe6ec47f96a5e18db16dc0)
  * Rename Kana indexing consts prefix into `KANA_`

Signed-off-by: JaeSang Yoo <jsyoo5b@gmail.com>

JSYoo5B committed:
refactor(automata): rename local vars in to_kana()

* rename `adj` to `diacrit` (diacritics offset)
  `adj` meant to calculate unicode offset for diacritics
* change `i`, `j` into `conso_idx`, `vowel_idx`
  they were temporal variable to index `kana_table`
  1. reading statement is far from writing statements
  2. single character named variables lower the readability
  the `i` used in buffer check (for loop in L209) were kept.
  (`i` in this context is temporal variable for index calculation)

Signed-off-by: JaeSang Yoo <jsyoo5b@gmail.com>

onting commented:
Your refactoring seems going well so far.

onting commented:
I guess _IDX infix in kana idxs can be omitted.
I feel we can know that the enums are table idxs clearly without the infix.
How do you think about it?
@JSYoo5B

JSYoo5B commented:
Well, I added infix `IDX` to mark the difference between the Unicode constants and indexing constants.
Maybe this idea was derived from `HANGUL` Unicode constants naming. (Hangul has each constant for each Jaso)

Anyway, your opinion is reasonable. I'll consider it. (and maybe apply)
Thx.

onting commented:
@JSYoo5B Can you give me a guarantee when this job would be finished?
#39 would goes with some code changes, and I want do it on your refactoring.
Thanks.

JSYoo5B commented:
@onting Sorry for the lazy updates.
I'll try to finish this refactor in the upcoming Sunday.

onting commented:
Good, #44 meight affect your affect.

JSYoo5B commented:
Rebased current refactoring into a recent bug-fixed version.

JSYoo5B committed:
refactor: extract HanjpBuffer from HanjpAutomata

* Although HanjpBuffer is only used in the HanjpAutomata,
  It has specific operations. may be better to split into specific file
* Change interface of `hanjp_buffer_peek()`
  It acts same as `top()` operation in general stack data structure.
  (actually it has minor difference from the stack's operation)

Signed-off-by: JaeSang Yoo <jsyoo5b@gmail.com>

JSYoo5B committed:
refactor: simplify hanjpbuffer codes

* Change magic nubmer (buffer size) into compile-time calculated const
* `hanjp_buffer_push()`: simplify code flow
* `hanjp_buffer_pop()`: refactor return buffer value

Signed-off-by: JaeSang Yoo <jsyoo5b@gmail.com>

JSYoo5B committed:
refactor: extract HanjpBuffer code block to func

* `hanjp_buffer_is_valid()`
  HanjpBuffer has only valid jamo
* `hanjp_buffer_copy_jamoes()`
  Copy valid jamo in HangulBuffer into GArray
* `hanjp_buffer_clear_filler()`
  Clears choseong filler or jungseong filler if it has
* `hanjp_buffer_aligh_jungseong()`
  Move second jungseong if first jungseong is empty

Signed-off-by: JaeSang Yoo <jsyoo5b@gmail.com>

JSYoo5B commented:
@onting I extracted `HanjpBuffer` from `HanjpAutomata`

They're strongly coupled. Most of the code blocks have their purpose. But they aren't abstracted enough. I'm trying to abstract them into more readable functions.

How do you think about these commits?

JSYoo5B commented:
The next remaining tasks are following:

1. Consume(eat) Choseong and Jungseong into Kana indexing => to internal functions
2. Reduce and divide combined Jungseong => to internal functions
3. Rename some short named variables (ex. `r`, `c`) and reused variables into proper name.

onting commented:
> @onting I extracted `HanjpBuffer` from `HanjpAutomata`
> 
> They're strongly coupled. Most of the code blocks have their purpose. But they aren't abstracted enough. I'm trying to abstract them into more readable functions.
> 
> How do you think about these commits?

I think it is good.
I had little worry about it when I first saw it because I wanted hide all buffer class operation except instanciation.
However, it can be safe with some protection.
So It feels okay.

onting commented:
> The next remaining tasks are following:
> 
> 1. Consume(eat) Choseong and Jungseong into Kana indexing => to internal functions
> 2. Reduce and divide combined Jungseong => to internal functions
> 3. Rename some short named variables (ex. `r`, `c`) and reused variables into proper name.

I think '2' is unnecessary.
I don't want add many hangul manipulating function explicitly.
Most of them are defined in libhangul, and I think additional operations are simple enough.
I think '3' is good.

JSYoo5B committed:
refactor: extract poping jamo from buffer to func

Signed-off-by: JaeSang Yoo <jsyoo5b@gmail.com>

JSYoo5B committed:
refactor: extract kana indexing into internal func

Signed-off-by: JaeSang Yoo <jsyoo5b@gmail.com>

JSYoo5B committed:
refactor: rename kana constants into upper letters

* KANA_TABLE and KANA_NN is constant variables.
  for better readability, rename into upper letters.
* KANA_SMALL_TU were represented by KANA_TABLE[T][U] - 1
  now, declare new constant KANA_SMALL_TU for better readability

Signed-off-by: JaeSang Yoo <jsyoo5b@gmail.com>

JSYoo5B committed:
refactor: rename internal vars

* `ch` used to represent hangul `jamo` and `kana`
  change each variable name by its purpose
* `val` used to represent combined jungseong when it exists.
  rename it to its purpose

Signed-off-by: JaeSang Yoo <jsyoo5b@gmail.com>

JSYoo5B committed:
refactor: extract to_kana code blocks to func

* Dividing jungseong code block into internal function
  it splits jungseong into two small jungseong for conversion
* Converting jongseong into Kana code block into internal function

Signed-off-by: JaeSang Yoo <jsyoo5b@gmail.com>

JSYoo5B commented:
@onting I finished refactoring related with `to_kana()` methods.

There are some other opinions about refactoring, but I'll write it as an issue. (This opinion may against your thought, so it could be a long discussion. Also, we need to prepare for releasing into stable version.)

onting commented on hanjpautomata.c:
```diff
@@ -120,7 +44,8 @@ static const gunichar kana_table[][5] = {
     {0x308F, 0x3090, 0x0000, 0x3091, 0x3092}  // WA
 };
-static const gunichar kana_nn = 0x3093;
+static const gunichar KANA_SMALL_TU = 0x3063;
```
Can you change it to KANA_SMALL_TSU?

JSYoo5B replied:
I also think `KANA_SMALL_TSU` is better. But, according to the Unicode, it has been defined as `SMALL_TU`. see following link: https://unicode-table.com/en/3063/

onting replied:
Alright.
Then, I also think tu is better.

onting reviewed (commented):
Your work seems good except bad indentation.
But it meight be fixed with some automated way.
